Bluetooth tester cleanup and BTP refactor

sjanc opened the pull request:
This PR cleanups and refactors BTP handling in Bluetooth tester application.
While working on it there is also multiple fixes for BTP handling like size validation, endianess, sync with BTP spec etc

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/53814

sjanc committed:
bluetooth: tester: Prefix tester BTP protocol defines

Make all BTP related defines and structures prefixed with BTP.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Move core service to separate file

There is no need to keep it in same file as BTP protocol handler.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Initialize GAP service synchronously

There is no need to initialize in asynchronous way as BTP is blocked
anyway until init is done.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Refactor BTP handling

This changes the way of BTP commmand handlers are called. Instead of
calling functions to handle service we now provide API for registering
callbacks per command. This allows to keep common checks (size etc)
in single place for (most) commands. Also provide common buffer for
reponses forcing callbacks to handle replay in common way.

In this commit only Core and GAP services are converted. This leaves
rest of services unfuntional (but code stil compiles correctly).
Other services are converted in following commits.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Convert BTP to use bt_addr_le_t

bt_addr_le_t matches BTP convention for addresses and it makes it
so much better to handle those directly from command structures.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Convert GATT service to new BTP API

This makes GATT service to use new BTP command handlers API.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Convert L2CAP service to new BTP API

This makes L2CAP service to use new BTP command handlers API.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Cleanup BTP definitions

Cleanup and unify names and conventions in BTP headers.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Fix IAS service registration

Make sure we don't send events if service was not registered.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Refactor VCP BTP

Split register functions, convert to new BTP API and fix some bugs
in BTP handlers.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Convert MESH service to new BTP API

This makes MESH service to use new BTP command handlers API.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Unify BTP index handling

We always have 0 index (and occasional NONE index) so there is no
point in validating those in each handler.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Convert PACS service to new BTP API

This makes PACS service to use new BTP command handlers API.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
@mkasenberg @KKopyscinski   FYI   you may need to re-sync when this PR gets in

sjanc commented:
#AutoPTS run zephyr

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1451936562 after 15:19:01.

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1451954894 after 15:30:18.

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1451954894 after 15:31:06.

sjanc committed:
bluetooth: tester: Keep BTP command in buffer for delayed response

This allows to avoid copying data for async callbacks.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
bluetooth: tester: Add explicit functions for event and response

This make is easier to verify if proper use of BTP is maintained.
tester_rsp and tester_rsp_full will be removed eventually when
BTP is fully async.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

Thalley commented:
@sjanc just FYI, you can easily run the compliance checks locally by running `/scripts/ci/check_compliance.py` :) That way you can verify it faster

Thalley commented on tests/bluetooth/tester/src/btp/btp_bap.h:
```diff
@@ -6,22 +6,20 @@
-struct pacs_read_supported_commands_rp {
+#define BTP_PACS_READ_SUPPORTED_COMMANDS			0x01
+struct btp_pacs_read_supported_commands_rp {
 	uint8_t data[0];
 } __packed;
 
-#define PACS_UPDATE_CHARACTERISTIC	0x02
-struct pacs_update_characteristic_cmd {
+#define BTP_PACS_CHARACTERISTIC_SINK_PAC			0x01
+#define BTP_PACS_CHARACTERISTIC_SOURCE_PAC			0x02
+#define BTP_PACS_CHARACTERISTIC_SINK_AUDIO_LOCATIONS		0x03
+#define BTP_PACS_CHARACTERISTIC_SOURCE_AUDIO_LOCATIONS		0x04
+#define BTP_PACS_CHARACTERISTIC_AVAILABLE_AUDIO_CONTEXTS	0x05
+#define BTP_PACS_CHARACTERISTIC_SUPPORTED_AUDIO_CONTEXTS	0x06
```
I appreciate that this is how it was done before, and how `hci.h` also does it, but we could also consider putting the commands into an `enum` as recommended by https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl

Thalley replied:
> as comparing value outside of enum specified range is not defined

I'm not sure I follow that; can you elaborate?

sjanc replied:
Those structs are used to parse incoming binary data from BTP, so lets say we have this:

enum BTP_PACS_CHR __packed {
   BTP_PACS_CHARACTERISTIC_SINK_PAC = 0x01,
   BTP_PACS_CHARACTERISTIC_SOURCE_PAC = 0x02,
}

struct btp_pacs_cmd {
  enum BTP_PACS_CHR chr;
}

and then we use it in handler cb to parse 1 byte buffer, where cmd[0] = 0x42
static uint8_t pacs_cmd(const void *cmd, uint16_t cmd_len,  void *rsp, uint16_t *rsp_len)
{
   const struct btp_pacs_cmd *cp = cmd;

   so now cp->chr has bianry value of 0x42 which is outside of defined enum range
}

sjanc replied:
but this is the case exactly I'm not liking about using enums for this,    doing switch on enum while value is outside of valid range

Thalley commented on tests/bluetooth/tester/src/btp/btp_ias.h:
```diff
@@ -10,7 +10,7 @@
 /* events */
-#define IAS_EV_OUT_ALERT_ACTION       0x80
-struct ias_alert_action_ev {
+#define BTP_IAS_EV_OUT_ALERT_ACTION	0x80
```
General comment: In Zephyr we use (or should use) spaces rather than tabs to indent after a non-tab character

Thalley commented on tests/bluetooth/tester/src/btp/btp_l2cap.h:
```diff
@@ -12,122 +12,116 @@
+#define BTP_L2CAP_TRANSPORT_LE			0x01
 
-#define L2CAP_CONNECTION_RESPONSE_SUCCESS		0x00
-#define L2CAP_CONNECTION_RESPONSE_INSUFF_AUTHEN		0x01
-#define L2CAP_CONNECTION_RESPONSE_INSUFF_AUTHOR		0x02
-#define L2CAP_CONNECTION_RESPONSE_INSUFF_ENC_KEY		0x03
+#define BTP_L2CAP_CONNECTION_RESPONSE_SUCCESS		0x00
+#define BTP_L2CAP_CONNECTION_RESPONSE_INSUFF_AUTHEN	0x01
+#define BTP_L2CAP_CONNECTION_RESPONSE_INSUFF_AUTHOR	0x02
+#define BTP_L2CAP_CONNECTION_RESPONSE_INSUFF_ENC_KEY	0x03
```
Consider aligning the values at the same column

Thalley commented on tests/bluetooth/tester/src/btp/btp_vcs.h:
```diff
@@ -8,14 +8,20 @@
 	uint8_t volume;
 } __packed;
+
+#define BTP_VCS_VOL_UP				0x03
+
+#define BTP_VCS_VOL_DOWN			0x04
+
+#define BTP_VCS_MUTE				0x05
+
+#define BTP_VCS_UNMUTE				0x06
```
```suggestion
#define BTP_VCS_VOL_UP				0x03
#define BTP_VCS_VOL_DOWN			0x04
#define BTP_VCS_MUTE				0x05
#define BTP_VCS_UNMUTE				0x06
```

Thalley commented on tests/bluetooth/tester/src/btp_vcp.c:
```diff
@@ -224,197 +218,160 @@ static struct bt_aics_cb aics_cb = {
+	},
+	{
+		.opcode = BTP_AICS_DESCRIPTION,
+		.expect_len = -1,
```
Can you explain the `-1` in length?

Thalley replied:
> yeap, I'll add comments in structure definition. -1 means that length is variable and is validated by handler itself

Suggest to use a `#define` for that then :) A named value would make it much more clear

Thalley reviewed (commented):
Approach seems good. A few nit-picky comments

sjanc commented on tests/bluetooth/tester/src/btp.c:
```diff
@@ -35,276 +35,100 @@ struct btp_buf {
+				  uint8_t status);
 
-static void register_service(uint8_t *data, uint16_t len)
+void tester_register_command_handlers(uint8_t service,
+				      const struct btp_handler *handlers,
+				      size_t num)
```
this is how BTP works,  ie services are registered (enabled) on demand

sjanc commented:
#AutoPTS run zephyr DIS GAP GATT SM L2CAP

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1453725667 after 02:48:48.

codecoup-tester commented:
AutoPTS Bot results:
Failed tests:
GAP     GAP/BROB/BCST/BV-03-C           INCONC
GAP     GAP/CONN/UCON/BV-06-C           FAIL
GAP     GAP/PRIV/CONN/BV-10-C           FAIL
GAP     GAP/SEC/AUT/BV-17-C             BTP ERROR
GAP     GAP/SEC/AUT/BV-19-C             BTP ERROR
GAP     GAP/SEC/AUT/BV-20-C             BTP TIMEOUT
GAP     GAP/SEC/AUT/BV-23-C             FAIL
GAP     GAP/SEC/AUT/BV-25-C             BTP ERROR - ERRATA CASE0072001
GAP     GAP/SEC/SEM/BV-62-C             INCONC - ERRATA TSE18349
GAP     GAP/SEC/SEM/BV-63-C             INCONC - ERRATA TSE18349
GAP     GAP/SEC/SEM/BV-64-C             INCONC - ERRATA TSE18349
GAP     GAP/SEC/SEM/BV-65-C             INCONC - ERRATA TSE18349
GAP     GAP/SEC/SEM/BV-66-C             INCONC - ERRATA TSE18349
GAP     GAP/SEC/SEM/BV-67-C             INCONC - ERRATA TSE18349
GATT    GATT/CL/GPM/BV-11-C             FAIL
GATT    GATT/SR/GAI/BV-01-C             BTP ERROR
GATT    GATT/SR/GAN/BV-01-C             BTP ERROR
GATT    GATT/SR/GAT/BV-01-C             BTP ERROR
L2CAP   L2CAP/ECFC/BV-32-C              FAIL - ERRATA Request ID 16410
L2CAP   L2CAP/ECFC/BV-43-C              FAIL - ERRATA TSE18351
L2CAP   L2CAP/ECFC/BV-44-C              INCONC - ERRATA TSE18351
L2CAP   L2CAP/TIM/BV-03-C               INCONC
SM      SM/CEN/KDU/BI-02-C              INCONC - ERRATA TSE19199
SM      SM/CEN/KDU/BI-03-C              INCONC - ERRATA TSE19199

sjanc commented:
#AutoPTS run zephyr AICS IAS PACS VCS VOCS

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1455085432 after 14:03:32.

sjanc commented:
#AutoPTS run zephyr MESH

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1455165526 after 19:26:26.

sjanc committed:
bluetooth: tester: Use define for variable length BTP commands

This makes code a bit easier to follow.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
#AutoPTS run zephyr AICS IAS PACS VCS VOCS

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1455693361 after 09:28:46.

sjanc committed:
bluetooth: tester: Use one-time advertising in GAP tests

This avoids potencial races where PTS is changing type of advertising
during tests.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
#AutoPTS run zephyr AICS IAS PACS VCS VOCS

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1456865311 after 20:54:41.

sjanc commented:
#AutoPTS run zephyr MESH

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55362#issuecomment-1458853576 after 21:51:01.

sjanc commented:
OK, so latest version seems to work OK in my local environment and I consider it ready to merged.
Any regressions that would show up can be handled afterwards

Thalley reviewed (approved):
I have no objections to these changes, LGTM
